src/main/java/frc/robot/subsystems/Vision.java

```diff
@@ -0,0 +1,88 @@
+/* ==================================================
+ * Author: Lucas Jacobs
+ *
+ * --------------------------------------------------
+ * Description:
+ *
+ * Contains functions needed to control the vision system.
+ * ================================================== */
+
+package frc.robot.subsystems;
+
+import org.opencv.core.Rect;
+import org.opencv.imgproc.Imgproc;
+
+import edu.wpi.first.cameraserver.CameraServer;
+import edu.wpi.first.cscore.UsbCamera;
+import edu.wpi.first.vision.VisionThread;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.Robot;
+
+public class Vision extends SubsystemBase {
+  // HARDWARE //
+
+  // PORTS //
+
+  // CONSTANTS //
+  final static int HEIGHT = 120;
+  final static int WIDTH = 160;
+  final static int FPS = 30;
+  final static double TURN_SPEED = 0.1;
+
+  // VARIABLES //
+  private VisionThread visionThread;
+  private double xCentre;
+  private Object imgLock;
+  private Robot robot;
+
+
+  public Vision(Robot robot) {
+    this.robot = robot;
+  }
+
+  // METHODS //
+
+  /* =====================================
+  Author: Lucas Jacobs
+
+  Desc:
+  This method initalizes the camera and the vison
+  thread so that frames from the camera can be accessed
+  ===================================== */
+  public void initCamera() {
+    // initalize camera
+    UsbCamera cam = CameraServer.startAutomaticCapture();
+
+    cam.setResolution(WIDTH, HEIGHT);
+    cam.setFPS(FPS);
+
+    // initalize vision thread
+    visionThread = new VisionThread(cam, new GripPipeline(), pipeline -> {
+      
+      if (!pipeline.filterContoursOutput().isEmpty()) {
+        Rect r = Imgproc.boundingRect(pipeline.filterContoursOutput().get(0)); //Gets bounding rectangle around the target
+        synchronized (imgLock) {
+          xCentre = r.x + (r.width / 2); //Set the sentre of the bounding rectangle
+        }
+      }
+    });
+
+    visionThread.start();
+  }
+
+  /* =====================================
+  Author: Lucas Jacobs
+
+  Desc:
+  This method makes the robot turn in place 
+  to face the target
+  ===================================== */
+  public void aimAtTarget() {
+    double xCentre;
+    synchronized (imgLock) {
+        xCentre = this.xCentre;
+    }
+    double turn = xCentre - (WIDTH / 2); // find difference between the target and where the robot is pointed
+    robot.drivetrain.moveMotors(0, turn * TURN_SPEED);
```

Comment:
When testing with low speeds, this could work to stop the robot from turning when perfectly aligned, but at higher turn speeds, due to the camera input delay, there is a good chance you will overshoot the target and end up swerving back the other way to correct it, leading to repeated oscillations(you will never get exactly on-target). A way to solve this is to add a threshold in the `turn` value for the alignment to be "good enough" and stop the robot from turning earlier instead of stopping when it is exactly on-target(to compensate for the camera lag). Another thing you can add (when you have time) is to make the turning speed proportional (Think of a PID control loop but just the proportional part), so that when the robot is more misaligned, it will turn faster, and when it gets closer to the target, it will turn slower. This is optional however since it would work well with just a fixed speed and threshold too.

Reply:
I added a threshold to the aiming. The speed is already proportional to the gap between the centre and the target. Do you think a full PID would be useful here, or is this good enough?

Reply:
Just using proportional control with a threshold is enough for the first competition. If you have time afterwards and you see that it can be improved, then you can maybe try implementing PI loop, probably don't need a full PID. Tuning the PID values will also be a pain so only work on it if it is critical and there is nothing else needed to be done.

Reply:
Okay, thanks.